### PR TITLE
Add inlined python udf and rust udf

### LINKED ISSUE #1833: Document: feat(udf): support inlined Rust UDF
This issue tracks the documentation update needed for the merged PR #14903.

Source PR URL: https://github.com/risingwavelabs/risingwave/pull/14903
Source PR Merged At: 2024-02-04T07:16:35Z

### LINKED ISSUE #1892: Document: feat(udf): support inlined python udf
This issue tracks the documentation update needed for the merged PR #15168.

Source PR URL: https://github.com/risingwavelabs/risingwave/pull/15168
Source PR Merged At: 2024-02-28T02:57:09Z

### PR DESCRIPTION
**Pull request by ShanlanLi**
## Info

- **Description**

  - [ as title ]

- **Notes**

  - [ Include any supplementary context or references here. ]

- **Related code PR**

  - https://github.com/risingwavelabs/risingwave/pull/15168
  - https://github.com/risingwavelabs/risingwave/pull/14903

- **Related doc issue**
  
  Resolves https://github.com/risingwavelabs/risingwave-docs/issues/1892#event-11996987063
 Resolves https://github.com/risingwavelabs/risingwave-docs/issues/1833#event-11997024509

## For reviewers

- **Preview**

  - [ Paste the preview link to the updated page(s) here. Edit this item after the preview site is ready. To find the updated pages, scroll down to locate and open the Amplify preview link and select the **dev** version of the documentation. ]

- **Key points**

  - [ Parts that may need revision or extra consideration. ]

## Before merging

- [ ] I have checked the doc site preview, and the updated parts look good.

- [ ] I have acquired the approval from the owner (and optionally the reviewers) of the code PR and at least one tech writer (`CharlieSYH`, `emile-00`, & `hengm3467`).

### CODE REVIEW

File: docs/sql/udf/user-defined-functions.md [hunk 1]
```diff
@@ -23,7 +23,7 @@ It is important to note that UDFs have lower execution efficiency compared to bu
 
 ## Types of UDFs in RisingWave
 
-At present, there are two ways to define your UDF in RisingWave. The first option is to use it as an external function, leveraging the capabilities of languages such as Python and Java to perform advanced data processing and analysis tasks. The second option is to use a SQL UDF. Currently, we support basic SQL UDFs with named and unnamed parameters, and we are actively working on enhancing the capabilities of SQL UDFs.
+At present, there are three ways to define your UDF. The first option is to use it as an external function, leveraging the capabilities of languages such as Python and Java to perform advanced data processing and analysis tasks. The second option is to use a embedded UDF, which runs internally in RisingWave. The third is to use a SQL UDF. Currently, we support basic SQL UDFs with named and unnamed parameters, and we are actively working on enhancing the capabilities of SQL UDFs.
```
Comment (wangrunji0408) [inline docs/sql/udf/user-defined-functions.md]:
```
leveraging ... analysis tasks
```
This description sounds a bit generic today. We may change it to:
```
which runs as a standalone service and provides maximum flexibility.
```
to emphasize the difference between external functions and embedded ones.


Besides, for SQL UDFs, my suggestion is:

```
The third is to use a SQL UDF, which allows for the same capabilities as regular SQL while offering a more concise way to express complex queries.
```

File: docs/sql/udf/user-defined-functions.md [hunk 2]
```diff
@@ -39,7 +39,12 @@ RisingWave supports creating UDFs with the following programming languages:
 
 - [Rust](/sql/udf/udf-rust.md)
 
+### Embedded UDFs
+
+- [Python](/sql/udf/udf-python.md)
```
Comment (wangrunji0408) [inline docs/sql/udf/user-defined-functions.md]:
```suggestion
- [Python](/sql/udf/udf-python-embedded.md)
```

File: docs/sql/udf/udf-rust-embedded.md
Comment (wangrunji0408) [inline docs/sql/udf/udf-rust-embedded.md:1]: This file can be merged into the original Rust UDF page. It only introduces a new way to define Rust UDFs without changing their existing functionality (unlike the new embedded Python UDF vs. the old external UDF). I'll push a commit to do this.

File: docs/sql/commands/sql-create-function.md [hunk 3]
```diff
@@ -16,6 +16,8 @@ You can define your own functions (including table functions) by some programmin
 
 The `CREATE FUNCTION` command is used to declare these UDFs. After that, you can use them in SQL queries like any built-in functions.
 
+Meanwhile, we also support inlining Python UDFs and Rust UDFs in the `CREATE FUNCTION` statement.
```
Comment (wangrunji0408) [inline docs/sql/commands/sql-create-function.md]: I don't think we need to mention embedded (inlined) UDFs here. Instead, let's update Line 11:

```
There are three ways to create UDFs in RisingWave: UDFs as external functions, embedded UDFs and SQL UDFs.
```

File: docs/sql/commands/sql-create-function.md [hunk 4]
```diff
@@ -75,6 +77,43 @@ CREATE FUNCTION gcd(a int, b int) RETURNS int LANGUAGE javascript AS $$
 $$;
 ```
 
+---
```
Comment (wangrunji0408) [inline docs/sql/commands/sql-create-function.md]: I'd like to add a title here.

```suggestion
## Embedded UDFs
```

File: docs/sql/commands/sql-create-function.md [hunk 5]
```diff
@@ -75,6 +77,43 @@ CREATE FUNCTION gcd(a int, b int) RETURNS int LANGUAGE javascript AS $$
...
+dev=> select series(5);
+ series 
+--------
+      0
+      1
+      2
+      3
+      4
+(5 rows)
+```
```
Comment (wangrunji0408) [inline docs/sql/commands/sql-create-function.md]:
```suggestion
```sql title="Inlined UDFs"
# Embedded Python UDF
create function gcd(a int, b int) returns int language python as $$
def gcd(a, b):
    while b != 0:
        a, b = b, a % b
    return a
$$;

# Embedded JavaScript UDF
create function gcd(a int, b int) returns int language javascript as $$
    while (b != 0) {
        let t = b;
        b = a % b;
        a = t;
    }
    return a;
$$;

# Embedded Rust UDF
create function gcd(int, int) returns int language rust as $$
    fn gcd(mut a: i32, mut b: i32) -> i32 {
        while b != 0 {
            let t = b;
            b = a % b;
            a = t;
        }
        a
    }
$$;
```
```

File: docs/sql/udf/udf-python-embedded.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+You can define embedded Python UDFs, which will be run in an embedded Python interpreter in RisingWave. The Python code is inlined in the [`CREATE FUNCTION`](/sql/commands/sql-create-function.md) statement.
+
+Here are some examples:
+
+- Scalar function example
+
+```sql title="Create function"
+CREATE FUNCTION gcd(a int, b int) RETURNS int LANGUAGE python AS $$
+DEF gcd(a, b):
```
Comment (wangrunji0408) [inline docs/sql/udf/udf-python-embedded.md]:
```suggestion
def gcd(a, b):
```

File: docs/sql/udf/udf-python-embedded.md
Comment (wangrunji0408) [inline docs/sql/udf/udf-python-embedded.md:1]: We need more details for this page. I'll add them later.

**Review by wangrunji0408 — COMMENTED** [review]
Generally LGTM. I will add more details to these pages shortly. Thanks!

File: docs/sql/commands/sql-create-function.md (hunk 3 above)
Comment (ShanlanLi) [reply docs/sql/commands/sql-create-function.md]: Thanks for your kind and patient modification of this page^_^. After I created this pr, I realized that I forgot to update the create function page, and found that I could improve the embedded udf pages better, as I found the existing javascript udf page is a good template. So I didn't make a review request right away and left to deal with other doc issues. That is why this page was a bit messy. I'll make changes according to your suggestions, and thanks for your further editing!

**Comment by wangrunji0408** [comment]
@ShanlanLi  Please help review my last 2 commits. Thanks!

**Comment by ShanlanLi** [comment]
> @ShanlanLi Please help review my last 2 commits. Thanks!

@wangrunji0408 Thanks! Made two modifications:
- remove :
```{
                      type: "doc",
                      id: "sql/udf/udf-rust-embedded",
                      label: "Rust",
                    },
```
from sidebar.js as embedded rust was merged into the existing page.

- in the create function page, move the example of javascript udf from the external udf part to embedded udf part according to new category, and add reference link for the example of each type of embedded udf.

**Review by wangrunji0408 — APPROVED** [review]
LGTM! Thanks

File: docs/sql/udf/udf-rust.md [hunk 7]
```diff
@@ -12,7 +12,77 @@ This article provides a step-by-step guide for defining Rust functions in Rising
...
+    }
+    #[function("key_value(varchar) -> struct KeyValue")]
+    fn key_value(kv: &str) -> Option<KeyValue<'_>> {
+        let (key, value) = kv.split_once('=')?;
+        Some(KeyValue { key, value })
+    }
+$$;
+```
+
+Firstly, you need to define a structure using `struct` and annotate it with `#[derive(StructType)]`. Its fields must match the struct type declared in the `CREATE FUNCTION` statement. Then, define the function and annotate it with the `#[function("...")]` macro. The string in the macro represents the SQL signature of the function, with the tail return type being `struct StructName`. You can refer to the documentation of [arrow-udf](https://docs.rs/arrow-udf/0.2.1/arrow_udf/attr.function.html) for the specific syntax.
```
Comment (emile-00) [inline docs/sql/udf/udf-rust.md]:
```suggestion
First, define a structure using `struct` and annotate it with `#[derive(StructType)]`. Its fields must match the struct type declared in the `CREATE FUNCTION` statement. Then, define the function and annotate it with the `#[function("...")]` macro. The string in the macro represents the SQL signature of the function, with the tail return type being `struct StructName`. For the specific syntax, see [arrow-udf](https://docs.rs/arrow-udf/0.2.1/arrow_udf/attr.function.html).
```

**Pull request merged**